### PR TITLE
K8SPXC-429 Add backupSource to restore resource

### PR DESCRIPTION
**Pull request by nonemax**
[K8SPXC-429](https://jira.percona.com/browse/K8SPXC-429)

### CODE REVIEW

File: e2e-tests/demand-backup/run [hunk 1]
```diff
@@ -6,6 +6,23 @@ set -o xtrace
 test_dir=$(realpath $(dirname $0))
 . ${test_dir}/../functions
 
+run_recovery_from_source() {
+    desc "aws restore with backup source"
+    stext=$(kubectl get PerconaXtraDBClusterBackup on-demand-backup-aws-s3 -o json | jq -c '.status')
+    content=$(jq --argjson s $stext '.spec.backupSource=$s' $test_dir/conf/restore-from-source-backup-aws-s3.json)
+    cat > $test_dir/conf/restore-from-source-backup-aws-s3.json <<<$content
+    kubectl_bin apply -f $test_dir/conf/restore-from-source-backup-aws-s3.json
```
Comment (delgod) [inline e2e-tests/demand-backup/run:14]: keep it simple
```suggestion
    stext=$(kubectl_bin get pxc-backup on-demand-backup-aws-s3 -o json | jq -c '.status')
    jq --argjson s "$stext" '.spec.backupSource=$s' $test_dir/conf/restore-from-source-backup-aws-s3.json \
        | kubectl apply -f -
    
```

File: e2e-tests/demand-backup/conf/restore-on-demand-backup-pvc-2.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: pxc.percona.com/v1
+kind: PerconaXtraDBClusterRestore
+metadata:
+  name: on-demand-backup-pvc-2
+spec:
+  pxcCluster: some-name
+  backupSource:
+    destination: pvc/some-name-xb-on-demand-backup-pvc
```
Comment (delgod) [inline e2e-tests/demand-backup/conf/restore-on-demand-backup-pvc-2.yaml:8]: I think there should be backupsource of second PVC
```suggestion
    destination: pvc/some-name-xb-on-demand-backup-pvc-2
```

File: e2e-tests/demand-backup/run [hunk 3]
```diff
@@ -21,9 +38,14 @@ main() {
     run_recovery_check "$cluster" "on-demand-backup-pvc"
     check_pvc_md5
 
+    run_backup         "$cluster" "on-demand-backup-pvc-2"
+    run_recovery_check "$cluster" "on-demand-backup-pvc-2"
```
Comment (delgod) [inline e2e-tests/demand-backup/run:42]: please rename
```suggestion
    run_backup         "$cluster" "on-demand-backup-pvc-backupsource"
    run_recovery_check "$cluster" "on-demand-backup-pvc-backupsource"
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/pxc/v1/pxc_prestore_types.go [hunk 4]
```diff
@@ -58,10 +59,6 @@ func (cr *PerconaXtraDBClusterRestore) CheckNsetDefaults() error {
 		return errors.New("pxcCluster can't be empty")
 	}
 
-	if cr.Spec.BackupName == "" {
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/apis/pxc/v1/pxc_prestore_types.go:61]: https://github.com/percona/percona-xtradb-cluster-operator/pull/556/commits/5ba5badc0990cc04a9cf0e209efaae66989c0247

File: e2e-tests/demand-backup/conf/restore-on-demand-backup-pvc-2.yaml (hunk 2 above)
Comment (nonemax) [reply e2e-tests/demand-backup/conf/restore-on-demand-backup-pvc-2.yaml]: https://github.com/percona/percona-xtradb-cluster-operator/pull/556/commits/edfe5858f06e192094e7703e3ba62b0afea5647e

File: e2e-tests/demand-backup/run (hunk 1 above)
Comment (nonemax) [reply e2e-tests/demand-backup/run]: https://github.com/percona/percona-xtradb-cluster-operator/pull/556/commits/edfe5858f06e192094e7703e3ba62b0afea5647e

File: e2e-tests/demand-backup/run (hunk 3 above)
Comment (nonemax) [reply e2e-tests/demand-backup/run]: https://github.com/percona/percona-xtradb-cluster-operator/pull/556/commits/edfe5858f06e192094e7703e3ba62b0afea5647e

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | failed |
| recreate | failed |
| big-data | passed |
| init-deploy | failed |
| storage | passed |
| self-healing | failed |
| limits | passed |
| scaling-proxysql | failed |
| monitoring | failed |
| restore-to-encrypted-cluster | passed |
| monitoring-2-0 | failed |
| self-healing-advanced | passed |
| upgrade | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| affinity | passed |
| upgrade-consistency | passed |
| security-context | passed |
| one-pod | passed |
| auto-tuning | failed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | failed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | failed |
| init-deploy | failed |
| storage | passed |
| big-data | passed |
| scaling | failed |
| self-healing | failed |
| limits | passed |
| scaling-proxysql | failed |
| restore-to-encrypted-cluster | passed |
| monitoring | failed |
| self-healing-advanced | passed |
| upgrade | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| affinity | passed |
| upgrade-consistency | passed |
| scheduled-backup | failed |
| security-context | passed |
| one-pod | passed |
| auto-tuning | failed |
| demand-backup-encrypted-with-tls | failed |
| proxysql-sidecar-res-limits | failed |
| users | failed |

File: e2e-tests/demand-backup/run [hunk 5]
```diff
@@ -6,6 +6,23 @@ set -o xtrace
 test_dir=$(realpath $(dirname $0))
 . ${test_dir}/../functions
 
+run_recovery_from_source() {
+    desc "aws restore with backup source"
+
+    stext=$(kubectl_bin get pxc-backup on-demand-backup-aws-s3 -o json | jq -c '.status')
+    jq --argjson s "$stext" '.spec.backupSource=$s' $test_dir/conf/restore-from-source-backup-aws-s3.json \
+    | kubectl_bin apply -f -
```
Comment (delgod) [inline e2e-tests/demand-backup/run:14]: please keep tabulation readable
```suggestion
        | kubectl_bin apply -f -
```

File: e2e-tests/functions [hunk 6]
```diff
@@ -262,6 +262,7 @@ run_mysql() {
 
     client_pod=$(get_client_pod)
     wait_pod $client_pod 1>&2
+    sleep 15
```
Comment (delgod) [inline e2e-tests/functions:265]: for what test it is needed?

File: e2e-tests/functions [hunk 7]
```diff
@@ -297,7 +298,7 @@ compare_mysql_cmd() {
     if [[ "$IMAGE_PXC" =~ 8\.0 ]] && [ -f ${test_dir}/compare/${command_id}${postfix}-80.sql ]; then
         expected_result=${test_dir}/compare/${command_id}${postfix}-80.sql
     fi
-
+    sleep 15
```
Comment (delgod) [inline e2e-tests/functions:301]: for what test it is needed?

File: pkg/apis/pxc/v1/pxc_prestore_types.go [hunk 8]
```diff
@@ -58,8 +59,8 @@ func (cr *PerconaXtraDBClusterRestore) CheckNsetDefaults() error {
...
-		return errors.New("backupName can't be empty")
+	if cr.Spec.BackupName == "" && cr.Spec.BackupSource == nil {
+		return errors.New("backupName or BackupSource can't be empty")
```
Comment (delgod) [inline pkg/apis/pxc/v1/pxc_prestore_types.go:63]:
```suggestion
		return errors.New("backupName and BackupSource can't be empty simultaneously")
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | failed |
| recreate | failed |
| big-data | passed |
| storage | passed |
| self-healing | failed |
| self-healing-advanced | passed |
| operator-self-healing | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| haproxy | passed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| upgrade | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| security-context | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| scheduled-backup | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| scaling | passed |
| scaling-proxysql | passed |
| recreate | passed |
| upgrade | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| security-context | passed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| big-data | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| scaling | passed |
| scaling-proxysql | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| security-context | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| demand-backup-encrypted-with-tls | failed |

File: pkg/controller/pxcrestore/controller.go [hunk 9]
```diff
@@ -212,6 +204,33 @@ func (r *ReconcilePerconaXtraDBClusterRestore) Reconcile(request reconcile.Reque
 	return rr, err
 }
 
+func (r *ReconcilePerconaXtraDBClusterRestore) getBackup(cr *api.PerconaXtraDBClusterRestore) (*api.PerconaXtraDBClusterBackup, error) {
+	bcp := &api.PerconaXtraDBClusterBackup{}
+
+	if cr.Spec.BackupSource != nil {
+		bcp.Name = cr.Name
```
Comment (qJkee) [inline pkg/controller/pxcrestore/controller.go:211]:
```
return &api.PerconaXtraDBClusterBackup{
                    Name: cr.Name,
                    ......
},nil
```

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| scaling-proxysql | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| security-context | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| storage | passed |
| scheduled-backup | failed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| big-data | passed |
| demand-backup-encrypted-with-tls | failed |

**Review by qJkee — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| storage | passed |
| scaling | passed |
| haproxy | passed |
| recreate | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| upgrade | passed |
| self-healing-advanced | passed |
| limits | passed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| monitoring | passed |
| security-context | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| scheduled-backup | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |
| demand-backup-encrypted-with-tls | failed |
| users | failed |

**Pull request merged**